File: laplace/driver.f90
! -*- Mode: Fortran90; -*-
!-----------------------------------------------------------------
! Daniel R. Reynolds
! SMU, Mathematics
! Math 6370
! 30 March 2009
!=================================================================


program Laplace2D
  !-----------------------------------------------------------------
  ! Description: 
  !   We set up the linear residual 
  !           L*u - f,
  !   where L is a standard 2D Laplace operator, and f and u are 
  !   given vectors.  These are decomposed using a 2D parallel 
  !   domain decomposition strategy.  We then call the routine 
  !   linresid2D to compute the linear residual above.
  !-----------------------------------------------------------------
  !======= Inclusions ===========
  use mpi
  implicit none

  !======= Interfaces ===========
  interface
     subroutine linresid2D(u,f,res,norm2,locN,locM,dx,dy,comm,ierr)
       integer, intent(in) :: locN, locM, comm
       integer, intent(out) :: ierr
       double precision, intent(in) :: dx, dy
       double precision, intent(in),  dimension(locN,locM) :: u, f
       double precision, intent(out), dimension(locN,locM) :: res
       double precision, intent(out) :: norm2
     end subroutine linresid2D
  end interface

  !======= Declarations =========
  double precision, allocatable :: u(:,:), f(:,:), res(:,:)
  integer :: N, M, locN, locM, i, j
  integer :: my_id, nprocs, px, py, ierr, comm
  integer :: buf(4), pdims(2), pcoords(2)
  logical :: periods(2)
  double precision :: dx, dy, xl, yl, x, y
  double precision :: stime, ftime, res2norm
  character*50 :: outfile
  namelist /inputs/ N, M, px, py

  !======= Internals ============

  ! intialize MPI
  call MPI_Init(ierr)
  if (ierr /= 0) then
     write(0,*) ' error in MPI_Init =',ierr
     stop
  endif
  call MPI_Comm_size(MPI_COMM_WORLD, nprocs, ierr)
  if (ierr /= 0) then
     write(0,*) ' error in MPI_Comm_size =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif
  call MPI_Comm_rank(MPI_COMM_WORLD, my_id, ierr)
  if (ierr /= 0) then
     write(0,*) ' error in MPI_Comm_rank =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif

  ! root gets problem information from input file
  if (my_id == 0) then
     open(100,file='input.txt',form='formatted')
     read(100,inputs)
     close(100)
     buf = (/ N, M, px, py /)
  end if

  ! root broadcasts information to other procs
  call MPI_Bcast(buf, 4, MPI_INTEGER, 0, MPI_COMM_WORLD, ierr)
  if (ierr /= 0) then
     print *, ' error in MPI_Bcast =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif

  ! extract values from buffer
  N  = buf(1)
  M  = buf(2)
  px = buf(3)
  py = buf(4)

  ! check that the processor layout and communicator size agree
  if (nprocs /= px*py) then
     if (my_id == 0) then
        write(0,*) ' error: incorrect processor layout'
        write(0,*) '     nprocs = ',nprocs
        write(0,*) '     px = ',px,',  py =',py
     endif
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif

  ! set up Cartesian communicator, comm
  pdims = (/ px, py /)
  periods = .false.
  call MPI_Cart_create(MPI_COMM_WORLD, 2, pdims, periods, .false., comm, ierr)
  if (ierr /= 0) then
     write(0,*) ' error in MPI_Cart_create =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif

  ! get this processor's new id and location from comm
  call MPI_Comm_rank(comm, my_id, ierr)
  if (ierr /= 0) then
     write(0,*) ' error in MPI_Comm_rank =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif
  call MPI_Cart_coords(comm, my_id, 2, pcoords, ierr)
  if (ierr /= 0) then
     write(0,*) ' error in MPI_Cart_coords =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif

  ! set local mesh sizes (last process in each dir takes up remainder)
  dx = 1.d0/N
  dy = 1.d0/M
  locN = N/px
  locM = M/py
  xl = dx*locN*pcoords(1)
  yl = dy*locM*pcoords(2)
  if (pcoords(1) == px-1) then
     if (locN*px /= N) then
        locN = N - locN*(px-1)
     end if
  end if
  if (pcoords(2) == py-1) then
     if (locM*py /= M) then
        locM = M - locM*(py-1)
     end if
  end if

  ! root node outputs some information to screen
  if (my_id == 0) then
       print *,'Laplace2D driver with ',nprocs,' processes'
       print *,'    global problem size = ',N,'  by ',M
       print *,'    local problem sizes = ',locN,'  by ',locM
  endif

  ! Allocate vector memory
  allocate(u(locN,locM), f(locN,locM), res(locN,locM), stat=ierr)
  if (ierr /= 0) then
     write(0,*) 'Laplace2D error: proc ',my_id, &
          ', failed to allocate array data'
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  end if

  ! start timer
  stime = MPI_Wtime()

  ! Set up u and f
  do j=1,locM
     y = yl + (j-1)*dy
     do i=1,locN
        x = xl + (i-1)*dx
        u(i,j) = exp(-2.d1*((x-0.5d0)**2 + (y-0.5d0)**2))
        f(i,j) = 1.d0
     end do
  end do
  
  ! Adjust u at domain boundary
  if (pcoords(1) == 0) then
     u(1,:) = 0.d0
  end if
  if (pcoords(1) == px-1) then
     u(locN,:) = 0.d0
  end if
  if (pcoords(2) == 0) then
     u(:,1) = 0.d0
  end if
  if (pcoords(2) == py-1) then
     u(:,locM) = 0.d0
  end if

  ! Wait until all procs have caught up
  call MPI_Barrier(comm, ierr)
  if (ierr /= 0) then
     write(0,*) 'Laplace2D error: failed MPI_Barrier with error = ',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  end if

  ! Compute linear residual
  call linresid2D(u, f, res, res2norm, locN, locM, dx, dy, comm, ierr)
  if (ierr /= 0) then
     write(0,*) ' error in linresid2D = ',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  end if
  if (my_id == 0) then
     print *,' residual: ||L*u-f||_2 = ',res2norm
  end if

  ! stop timer
  ftime = MPI_Wtime()
  if (my_id == 0) then
     print *,' runtime = ',ftime-stime
  end if

  
  ! output residual to file(s)
  !   root outputs problem/parallelism information
  if (my_id == 0) then
     open(101,file='resid.txt')
     write(101,*) N, M, px, py
     close(101)
  endif
  write(outfile,'(5Hresid,f4.3)') float(my_id)/1000.0
  open(102,file=outfile)
  write(102,*) locN
  write(102,*) locM
  write(102,*) pcoords(1)
  write(102,*) pcoords(2)
  do j=1,locM
     do i=1,locN
        write(102,*) res(i,j)
     end do
  end do
  close(102)
  
  ! Free allocated memory
  deallocate(u,f,res)

  ! finalize MPI
  call MPI_Finalize(ierr)

end program Laplace2D
!=================================================================

File: laplace/driver.cpp
/* Daniel R. Reynolds
   SMU Mathematics
   Math 4370 / 6370 */

// Inclusions
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>
#include "mpi.h"
#include "laplace2d.hpp"


/*   We set up the linear residual
             L*u - f,
     where L is a standard 2D Laplace operator, and f and u are
     given vectors.  These are decomposed using a 2D parallel
     domain decomposition strategy.  We then call the routine
     linresid2D to compute the linear residual above. */
int main(int argc, char* argv[]) {

  // intialize MPI
  int ierr, numprocs, myid;
  ierr = MPI_Init(&argc, &argv);
  if (ierr != MPI_SUCCESS) {
    fprintf(stderr," error in MPI_Init = %i\n",ierr);
    return 1;
  }
  ierr = MPI_Comm_size(MPI_COMM_WORLD, &numprocs);
  if (ierr != MPI_SUCCESS) {
    fprintf(stderr," error in MPI_Comm_size = %i\n",ierr);
    MPI_Abort(MPI_COMM_WORLD, 1);
  }
  ierr = MPI_Comm_rank(MPI_COMM_WORLD, &myid);
  if (ierr != MPI_SUCCESS) {
    fprintf(stderr," error in MPI_Comm_rank = %i\n",ierr);
    MPI_Abort(MPI_COMM_WORLD, 1);
  }

  /* read problem parameters from input file (should be in this order):
         N - number of nodes in x-direction
	 M - number of nodes in y-direction
	px - number of procs in x-direction
	py - number of procs in y-direction */
  int N, M, px, py, buf[4];
  FILE *FID;
  if (myid == 0) {
    FID=fopen("input.txt","r");
    ierr = fscanf(FID," &inputs\n");
    ierr = fscanf(FID,"  N = %i,\n", &N);
    ierr = fscanf(FID,"  M = %i,\n", &M);
    ierr = fscanf(FID,"  px = %i,\n", &px);
    ierr = fscanf(FID,"  py = %i,\n", &py);
    fclose(FID);

    // fill send buffer
    buf[0] = N;
    buf[1] = M;
    buf[2] = px;
    buf[3] = py;
  }

  // broadcast data to all procs
  ierr = MPI_Bcast(&buf, 4, MPI_INT, 0, MPI_COMM_WORLD);
  if (ierr != MPI_SUCCESS) {
    fprintf(stderr," error in MPI_Bcast = %i\n",ierr);
    MPI_Abort(MPI_COMM_WORLD, 1);
  }

  // extract values from buffer
  N  = buf[0];
  M  = buf[1];
  px = buf[2];
  py = buf[3];

  // check that the processor layout and communicator size agree
  if (numprocs != px*py) {
    if (myid == 0) {
      fprintf(stderr," error: incorrect processor layout\n");
      fprintf(stderr,"     numprocs = %i\n",numprocs);
      fprintf(stderr,"     px = %i,  py = %i\n",px,py);
    }
    MPI_Abort(MPI_COMM_WORLD, 1);
  }

  // set up Cartesian communicator, comm
  int pdims[2], periods[2];
  MPI_Comm comm;
  pdims[0] = px;
  pdims[1] = py;
  periods[0] = 0;
  periods[1] = 0;
  ierr = MPI_Cart_create(MPI_COMM_WORLD, 2, pdims, periods, 0, &comm);
  if (ierr != MPI_SUCCESS) {
    fprintf(stderr," error in MPI_Cart_create = %i\n",ierr);
    MPI_Abort(MPI_COMM_WORLD, 1);
  }

  // get this processor's new id and location from comm
  int pcoords[2];
  ierr = MPI_Comm_rank(comm, &myid);
  if (ierr != MPI_SUCCESS) {
    fprintf(stderr," error in MPI_Comm_rank = %i\n",ierr);
    MPI_Abort(comm, 1);
  }
  ierr = MPI_Cart_coords(comm, myid, 2, pcoords);
  if (ierr != MPI_SUCCESS) {
    fprintf(stderr," error in MPI_Cart_coords = %i\n",ierr);
    MPI_Abort(comm, 1);
  }

  // set local mesh sizes (last process in each dir takes up remainder)
  const double dx = 1.0/(N-1);
  const double dy = 1.0/(M-1);
  int locN = N/px;
  int locM = M/py;
  const double xl = dx*locN*pcoords[0];
  const double yl = dy*locM*pcoords[1];
  if (pcoords[0] == px-1)
    if (locN*px != N)
      locN = N - locN*(px-1);
  if (pcoords[1] == py-1)
    if (locM*py != M)
      locM = M - locM*(py-1);

  // root node outputs some information to screen
  if (myid == 0) {
    printf("Laplace2D driver with %i processes\n", numprocs);
    printf("    global problem size = %i by %i\n", N, M);
    printf("    local problem sizes = %i by %i\n", locN, locM);
  }

  // allocate vector memory
  double *u   = new double[locN*locM];
  double *f   = new double[locN*locM];
  double *res = new double[locN*locM];

  // start timer
  double stime = MPI_Wtime();

  // set up u and f
  int i, j;
  for (j=0; j<locM; j++) {
    double y = yl + j*dy;
    for (i=0; i<locN; i++) {
      double x = xl + i*dx;
      u[idx(i,j,locN)] = exp(-20.0*((x-0.5)*(x-0.5) + (y-0.5)*(y-0.5)));
      f[idx(i,j,locN)] = 1.0;
    }
  }

  // adjust u at domain boundary
  if (pcoords[0] == 0)
    for (j=0; j<locM; j++)
      u[idx(0,j,locN)] = 0.0;
  if (pcoords[0] == px-1)
    for (j=0; j<locM; j++)
      u[idx(locN-1,j,locN)] = 0.0;
  if (pcoords[1] == 0)
    for (i=0; i<locN; i++)
      u[idx(i,0,locN)] = 0.0;
  if (pcoords[1] == py-1)
    for (i=0; i<locN; i++)
      u[idx(i,locM-1,locN)] = 0.0;

  // wait until all procs have caught up
  ierr = MPI_Barrier(comm);
  if (ierr != MPI_SUCCESS) {
    fprintf(stderr," error in MPI_Barrier = %i\n",ierr);
    ierr = MPI_Abort(comm, 1);
  }

  /* compute linear residual */
  double res2norm;
  ierr = linresid2D(u, f, res, res2norm, locN, locM, dx, dy, comm);
  if (ierr != MPI_SUCCESS) {
    printf(" error in linresid2D = %i\n",ierr);
    ierr = MPI_Abort(comm, 1);
  }
  if (myid == 0)
    printf(" residual: ||L*u-f||_2 = %g\n",res2norm);

  // stop timer
  double ftime = MPI_Wtime();
  if (myid == 0)
    printf(" runtime = %g\n",ftime-stime);

  // output residual to file(s)
  if (myid == 0) {
    FID = fopen("resid.txt","w");
    fprintf(FID, "%i %i %i %i\n", N, M, px, py);
    fclose(FID);
  }
  char outname[100];
  sprintf(outname, "resid.%03i", myid);
  FID = fopen(outname,"w");
  fprintf(FID, "%i\n", locN);
  fprintf(FID, "%i\n", locM);
  fprintf(FID, "%i\n", pcoords[0]);
  fprintf(FID, "%i\n", pcoords[1]);
  for (j=0; j<locM; j++)
    for (i=0; i<locN; i++)
      fprintf(FID, "%.16e\n",res[idx(i,j,locN)]);
  fclose(FID);

  // free memory
  delete[] u;
  delete[] f;
  delete[] res;

  // finalize MPI
  MPI_Finalize();

} // end main